Better safety reasoning about CPU-local variables

lrh2000 opened the pull request:
I'm trying to get avoid of some "fake" safety reasoning. For example:
https://github.com/asterinas/asterinas/blob/b9ce3e64ad57f336b22ae39daaae618ad12c343c/ostd/src/arch/x86/cpu/local.rs#L49-L52
- `SAFETY: GS has a valid base`, but why? **(Problem 1)**

https://github.com/asterinas/asterinas/blob/b9ce3e64ad57f336b22ae39daaae618ad12c343c/ostd/src/cpu/local/cpu_local.rs#L152-L157
- `SAFETY: The storage is initialized`, but why? **(Problem 2)**

https://github.com/asterinas/asterinas/blob/b9ce3e64ad57f336b22ae39daaae618ad12c343c/ostd/src/cpu/mod.rs#L85-L87
- `SAFETY: The number of CPUs is initialized`, but why? **(Problem 3)**

I found myself unable to reasonably answer the above "but why" questions. Of course, I know that these safety conditions are perfectly met _after our kernel has successfully booted_, but the problem is _who ensures that these methods can't be called during boot time?_ They're exposed to safe APIs, so it's just too easy to call them at boot time without realizing that it introduces UB.

This got me thinking about whether it would be possible to fix the problems. In fact, I found that all three problems are fixable: some tricks are involved, but overall I think that fixing them results in a better state for doing proper safety reasoning.

- To fix **Problem 1**, we can simply introduce a new global invariant: the GS base address points to CPU-local memory. Saying it's a global invariant means that the initialization should be done before entering the Rust code. This seems reasonable for the following reasons:
  - Setting up the base address is highly architecture dependent, so using assembly code to initialize it before entering the Rust code seems fine.
  - We have already used `early_init_bsp_local_base` to set the base address early in the boot process. So moving the initialization to an even earlier stage of the boot process seems to be fine as well.
https://github.com/asterinas/asterinas/blob/b9ce3e64ad57f336b22ae39daaae618ad12c343c/ostd/src/lib.rs#L80-L81

For details, see commit "Turn `GS.base` validity into a global invariant".

- To fix **Problem 3**, we notice at the point of doing `CPU_LOCAL_STORAGES.get_unchecked()` that we must have a non-BSP `CpuId`. We can do the following to ensure that if a non-BSP `CpuId` can be constructed, then `CPU_LOCAL_STORAGES` must have been initialized.
  - First, we initialize `NUM_CPUS` to one at the beginning of the boot process, so that a non-BSP `CpuId` cannot be constructed.
  - Second, we initialize `CPU_LOCAL_STORAGES` (according to the actual number of CPUs).
  - Third, we update `NUM_CPUS` to the actual number of CPUs. Now a non-BSP `CpuId` can be constructed.

For details, see commit "Fix safety reasoning about `get_on_cpu`" (Note that it also fixes **Problem 2**).

(Looks like #1783 also touches the related code. This PR may need to be rebase after it is merged).

junyang-zh commented:
This is an early response to this PR and thanks for making suggestions about CPU-local variables. I am highly in favor of your idea, especially about the timing of initializing `GS`, which has introduced problems when I am improving the MM system.

However I believe problem 1, 2 and 3 are resolved by #1783 ([this commit](https://github.com/asterinas/asterinas/pull/1783/commits/15e1ac5bced5ad18fb7e7495c2aa6d3fdf79ab59)). I am sorry that #1783 is rather a huge PR and I didn't bring forward much attention on it's boot part in that PR. And I guess we are actually doing repetitive works this time (which proves that it is worth doing).

lrh2000 commented:
@junyang-zh Thanks for your quick response.

> However I believe problem 1, 2 and 3 are resolved by #1783 ([this commit](https://github.com/asterinas/asterinas/pull/1783/commits/15e1ac5bced5ad18fb7e7495c2aa6d3fdf79ab59)). I am sorry that #1783 is rather a huge PR and I didn't bring forward much attention on it's boot part in that PR. And I guess we are actually doing repetitive works this time (which proves that it is worth doing).

I noticed the problems at the end of 2024 and did most of the work around the new year 2025. So it is unfortunate (but possible) that we are doing the same work at the same time. I will look at your commit when I have some time. 

But I don't think your commit covers everything in this PR? Let's merge your PR first, then I'll rebase this PR to see if anything is missing and still needs to be done. (For example, this PR also adds a bunch of missing safety documentation. I didn't mention those things in the PR description).

junyang-zh commented:
> But I don't think your commit covers everything in this PR?

Ah, yes. The CPU-local portion of #1783 is not as carefully written as yours. And I'll appreciate if we merge #1783 first.

There are some design differences. #1783 uses an early frame allocator to allow CPU-local memory to be copied really early, so that the BSP can reuse the section. And I don't know how this PR addresses the problem that CPU-local memory would be accessed before booting APs (because disabling preempt would touch the section before copying them to the per-AP pages, and we deem to touch it since APs can start quite late).

lrh2000 commented:
> And I don't know how this PR addresses the problem that CPU-local memory would be accessed before booting APs (because disabling preempt would touch the section before copying them to the per-AP pages, and we deem to touch it since APs can start quite late).

IMO, this is not a problem introduced by this PR, and this PR will not address this problem at all.

In practice, we can allow BSP to touch "scalar" CPU-local variables (through `CpuLocalCell`), but it should be responsible that the mutated values are copied to APs as well.

lrh2000 commented:
Rebased. I think this PR should still benefit the safety reasoning.

This PR also helps to enable the `unsafe_op_in_unsafe_fn` lint since I've added quite a number of missed `unsafe` blocks and `SAFETY` comments when accessing the CPU local variables.

junyang-zh commented on ostd/src/boot/smp.rs:
```diff
@@ -50,57 +62,73 @@ static AP_LATE_ENTRY: Once<fn()> = Once::new();
+    log::info!("Booting {} processors", num_cpus - 1);
+
+    // We support up to 1024 APs.
+    assert!(num_cpus < 1024, "There are too many CPUs");
```
We used to allocate only two pages for the per-AP info array. Now I see you allocate pages according to the number of APs. So I think we can remove the assert now? As no code elsewhere have this assumption (`CpuSet` is implemented by a `SmallVec` so it can be dynamically off-stack if number of CPUs are large).

junyang-zh commented on ostd/src/cpu/mod.rs:
```diff
@@ -50,46 +49,62 @@ impl TryFrom<usize> for CpuId {
 
+cpu_local_cell! {
+    /// The ID of the current CPU.
+    static CURRENT_CPU: u32 = 0;
```
The initial value is zero while the `current_cpu_racy` function at line 136 asserts if it is initialized by checking if it equals to `u32::MAX`. I think we could make this value initialized as `u32::MAX` as we previously did.

lrh2000 replied:
The reason I change the initialization value from `u32::MAX` to `0` is because if it is initialized as `u32::MAX`, we will have a period of time (during the early boot stage) where creating `CpuId(current_cpu()) == CpuId(u32::MAX)` is legal, breaking the type invariant of the `CpuId` type.

This is not really important because the period in question is quite short. But if we can avoid it, it should be better.

On second thought, I think adding another boolean variable (`IS_CURRENT_CPU_INITED`) to keep track of whether `CURRENT_CPU` is set sounds good, I'll give it a try.

junyang-zh reviewed (commented):
Good work! I have only two minor comments.
